Add Cassandra object mapper

zwennesm opened the pull request:
This PR adds the datastax mapping manager to the `vertx-cassandra-client`.
Solves https://github.com/vert-x3/vertx-cassandra-client/issues/18

Example code:
```
@Table(keyspace = "test", name = "users")
class User {
  @PartitionKey String name;
  ..rest of POJO
}

VertxMappingManager manager = VertxMappingManager.create(cassandraClient);
VertxMapper<User> mapper = manager.mapper(User.class, vertx);

mapper.save(new User("john", hander -> {}));
```

Let me know how you guys feel about it.

vietj commented:
you are just killing one item of the 2019 vertx roadmap :-) - good job

vietj commented:
removing reviewers as it's not yet done

zwennesm commented:
Cool! Good to know that it was on the roadmap for this year. I've added the last bits (examples, copyright), so i guess it's ready for review @vietj

vietj commented:
@tsegismont will take care of you

Sammers21 commented on src/main/java/examples/CassandraClientExamples.java:
```diff
@@ -192,4 +196,31 @@ public void batching(CassandraClient cassandraClient) {
     });
   }
+
+  public void mapper(CassandraClient cassandraClient, Vertx vertx) {
```
You can refer to the example from a related documentation section. Look at `vertx-cassandra-client/src/main/asciidoc/index.adoc` line 57 for an example

zwennesm replied:
Docs have been updated.

Sammers21 commented on src/main/java/io/vertx/cassandra/mapper/VertxMapper.java:
```diff
@@ -0,0 +1,61 @@
+ *
+ * @author Martijn Zwennes
+ */
+public interface VertxMapper<T> {
```
`@VertxGen` here and the same for `VertxMappingManager`

Sammers21 replied:
These methods:
```java
VertxMappingManager#getMappingManager
VertxMapper#getMapper
VertxMapper#delete(Handler<AsyncResult<Void>> handler, Object... primaryKey)
VertxMapper#get(Handler<AsyncResult<Void>> handler, Object... primaryKey)
```
should be marked with  `@GenIgnore(GenIgnore.PERMITTED_TYPE)`

zwennesm replied:
`@VertxGen` has been added to both interfaces and the `@GenIgnore` annotation has been given to the types that needed it.

My apologies, i'm not familiar with `@VertxGen`

zwennesm commented on src/test/java/io/vertx/cassandra/MapperTest.java:
```diff
@@ -0,0 +1,98 @@
+    getHandler.setHandler(ar -> {
+      if (ar.succeeded()) {
+        context.assertEquals(test.name, ar.result().name);
+        context.assertEquals(test.age, ar.result().age);
```
I new something was missing in the tests, good find! It has been  added.

Sammers21 commented on src/test/java/io/vertx/cassandra/MapperTest.java:
```diff
@@ -0,0 +1,101 @@
+public class MapperTest extends CassandraServiceBase {
+
+  @Test
+  public void mappingManagerInitialize(TestContext context) {
```
We need to use **_context.async()_** in the test as well.

Sammers21 commented on src/main/java/io/vertx/cassandra/mapper/VertxMapper.java:
```diff
@@ -0,0 +1,68 @@
+   * @param handler result handler
+   */
+  @GenIgnore(GenIgnore.PERMITTED_TYPE)
+  void delete(T entity, Handler<AsyncResult<Void>> handler);
```
you don't need to annotate the method with `@GenIgnore(GenIgnore.PERMITTED_TYPE)`.

Sammers21 commented on src/main/java/io/vertx/cassandra/mapper/VertxMapper.java:
```diff
@@ -0,0 +1,68 @@
+import io.vertx.core.Handler;
+
+/**
+ * A vert.x wrapped datastax Mapper {com.datastax.driver.mapping.Mapper}
```
vert.x -> Vert.x

tsegismont commented on src/main/java/io/vertx/cassandra/mapper/VertxMapper.java:
```diff
@@ -0,0 +1,67 @@
+   * @param primaryKey primary key(s) used to find row(s) to delete
+   */
+  @GenIgnore(GenIgnore.PERMITTED_TYPE)
+  void delete(Handler<AsyncResult<Void>> handler, Object... primaryKey);
```
This won't work with codegen because it expects the handler of async result as last parameter. Can you update the method to:

```java
void delete(List<Object> primaryKeys , Handler<AsyncResult<Void>> handler);
```

Then I belive you don't need @GenIgnore any more

zwennesm replied:
When doing it this way, it seems that i have to `@GenIgnore` one of the delete methods because JS can't see the difference between `T entity` and `List<Object>`.

This is the error message i get when removing `@GenIgnore` at both delete methods:
```
SEVERE: io.vertx.codegen.CodeGenProcessor - Could not generate model for io.vertx.cassandra.VertxMapper: Failed to generate because it would be impossible in target language JS at runtime to resolve which of the following overloaded methods to call in the Java API: delete(T entity, io.vertx.core.Handler<io.vertx.core.AsyncResult<java.lang.Void>> handler) and delete(java.util.List<java.lang.Object> primaryKeys, io.vertx.core.Handler<io.vertx.core.AsyncResult<java.lang.Void>> handler) 
io.vertx.codegen.GenException: Failed to generate because it would be impossible in target language JS at runtime to resolve which of the following overloaded methods to call in the Java API: delete(T entity, io.vertx.core.Handler<io.vertx.core.AsyncResult<java.lang.Void>> handler) and delete(java.util.List<java.lang.Object> primaryKeys, io.vertx.core.Handler<io.vertx.core.AsyncResult<java.lang.Void>> handler)
```

tsegismont replied:
I see. Keep the `@GenIgnore` then, but switch the parameters. Otherwise we won't get proper rxified and coroutines API generation.

zwennesm replied:
I've switched the function parameters so that all functions have the handler in the end. Seems that this is also nice for `Kotlin` users because then they can use something like: 
get(primaryKeys) {
  //handler
}

tsegismont replied:
:+1:

tsegismont commented on src/main/java/io/vertx/cassandra/mapper/VertxMapper.java:
```diff
@@ -0,0 +1,67 @@
+   * @param primaryKey primary key(s) used to retrieve row(s)
+   */
+  @GenIgnore(GenIgnore.PERMITTED_TYPE)
+  void get(Handler<AsyncResult<T>> handler, Object... primaryKey);
```
Same here as with `delete`

tsegismont commented on src/main/java/io/vertx/cassandra/CassandraClient.java:
```diff
@@ -225,4 +226,12 @@ static CassandraClient createShared(Vertx vertx, String datasourceName, Cassandr
+   * @return the Datastax object mapping manager
+   */
+  @GenIgnore(GenIgnore.PERMITTED_TYPE)
+  VertxMappingManager getMappingManager();
```
Instead of retrieving an instance here, can you add a `create(CassandraClient)` method to `VertxMappingManager`? Then we'll be able to make the dependency on the Cassandra mapping manager optional.

zwennesm replied:
Creating a `MappingManager` requires a `Session` instance to be passed as a parameter. Currently the `Session` is hidden in the `CassandraHolder`, therefore it's not possible to create a `MappingManager` from a `CassandraSession`. 

So i went for the option of adding it to the `CassandraClient`. Do you have any suggestions to do it differently?

tsegismont replied:
In the mapping manager implementation, you can cast the cassandra client to its implementation class (we do this very often in Vert.x core and modules):

```java
static void create(CassandraClient client) {
  CassandraClientImpl clientImpl =  (CassandraClientImpl) client;
  // get the session from client impl
}
```

It is fine to add a method in the client implementation to return the current session.

zwennesm replied:
I added a package private getter for the `Session` in the `CassandraClientImpl` which can now be retrieved with your cast suggestion. The casting now happens in a static function called `create` which is located in the `VertxMappingManager`. 

On a side-note: I really appreciate the help. It's a good way to learn more about Vert.x

tsegismont replied:
Thanks!

tsegismont commented on src/main/java/io/vertx/cassandra/impl/Util.java:
```diff
@@ -29,12 +29,12 @@
  * @author Thomas Segismont
  */
-class Util {
+public class Util {
```
Please move the mapping manager interface to `io.vertx.cassandra` package and implementation to `io.vertx.cassandra.impl`. It's not very useful to have it separately and `Util` class can be left as-is

zwennesm replied:
I moved both the `MappingManager` and the `Mapper` to the `io.vertx.cassandra` package

tsegismont commented on src/main/java/io/vertx/cassandra/mapper/VertxMapper.java:
```diff
@@ -0,0 +1,67 @@
+import io.vertx.core.Handler;
+
+/**
+ * A Vert.x wrapped datastax Mapper {com.datastax.driver.mapping.Mapper}
```
All Javadoc first sentences must end with a dot. Also, to generate a correct link you need;

```
{@link com.datastax.driver.mapping.Mapper}
```

zwennesm replied:
Alright. I've added them.

tsegismont commented on src/main/java/io/vertx/cassandra/mapper/impl/DefaultVertxMapper.java:
```diff
@@ -0,0 +1,74 @@
+/**
+ * @author Martijn Zwennes
+ */
+public class DefaultVertxMapper<T> implements VertxMapper<T> {
```
By convention, we use the `Impl` suffix instead of the `Default` prefix

tsegismont commented:
@zwennesm I will take another look asap

zwennesm commented:
I saw that we have another PR that got merged with added some conflicts. Is there something that i can do to help to get this PR going?

tsegismont commented:
@zwennesm I had planned to review your PR again earlier this week, but got sick and so afk for a few days. Your proposal is on the 3.7 roadmap so rest assured it will hit master.

zwennesm commented:
@tsegismont I've updated the docs and changed the dates in the copyright statements, and I just signed the Eclipse Contributor Agreement. 

Thanks for all the help regarding feedback and merging, I really appreciate it.

zwennesm commented:
@tsegismont I've updated the docs and changed the dates in the copyright statements, squashed the commits, and I just signed the Eclipse Contributor Agreement.

Thanks for all the help regarding feedback and merging, I really appreciate it.

tsegismont commented:
@zwennesm I could not find your email address in the ECA verification tool. Can you signoff the commit with the email address used for signing the ECA?

zwennesm commented:
@tsegismont it's signed off with the correct email now

tsegismont commented:
@zwennesm can you squash again please? Thanks

zwennesm commented:
@tsegismont :+1:

tsegismont commented:
Closed via d21234d

tsegismont commented:
@zwennesm @Sammers21 I've rebased the PR on master and reworked it a bit:

- we no longer have `connect` method and `Session` might not be ready when `MappingManager`/`Mapper` are created (very likely when initializing a verticle)
- class names should be aligned with existing code (some had `Vertx` prefix)
- test keyspace/table should be created only in the corresponding test
- dependency on Datastax mapping manager should be optional
